src/queues/EventRedactionQueue.ts

```diff
@@ -0,0 +1,139 @@
+ */
+export class EventRedactionQueue {
+    private toRedact: Map<string, QueuedRedaction[]> = new Map<string, QueuedRedaction[]>();
+
+    /**
+     * Test whether the redaction is already present in the queue.
+     * @param redaction a QueuedRedaction.
+     * @returns True if the queue already has the redaction, false otherwise.
+     */
+    public has(redaction: QueuedRedaction): boolean {
+        return !!this.toRedact.get(redaction.roomId)?.find(r => r.redactionEqual(redaction));
+    }
+
+    /**
+     * Adds a QueuedRedaction to the queue to be processed when process is called.
+     * @param redaction A QueuedRedaction to await processing
+     * @returns A boolean that is true if the redaction was added to the queue and false if it was duplicated.
+     */
+    public add(redaction: QueuedRedaction): boolean {
+        if (this.has(redaction)) {
+            return false;
+        } else {
+            let entry = this.toRedact.get(redaction.roomId);
+            if (entry) {
+                entry.push(redaction);
+            } else {
+                this.toRedact.set(redaction.roomId, [redaction]);
+            }return true;
+        }
+    }
+
+    /**
+     * Process the redaction queue, carrying out the action of each QueuedRedaction in sequence.
+     * @param client The matrix client to use for processing redactions.
+     * @param limitToRoomId If the roomId is provided, only redactions for that room will be processed.
+     * @returns A description of any errors encountered by each QueuedRedaction that was processed.
+     */
+    public async process(client: MatrixClient, limitToRoomId?: string): Promise<RoomUpdateError[]> {
+        const errors: RoomUpdateError[] = [];
+        const redact = async (currentBatch: QueuedRedaction[]) => {
+            for (const redaction of currentBatch) {
+                try {
+                    await redaction.redact(client);
+                } catch (e) {
+                    let roomError: RoomUpdateError;
+                    if (e.roomId && e.errorMessage && e.errorKind) {
+                        roomError = e;
+                    } else {
+                        const message = e.message || (e.body ? e.body.error : '<no message>');
+                        roomError = {
+                            roomId: redaction.roomId,
+                            errorMessage: message,
+                            errorKind: ERROR_KIND_FATAL,
+                        };
+                    }
+                    errors.push(roomError);
+                }
+            }
+        }
+        if (limitToRoomId) {
+            // There might not actually be any queued redactions for this room.
+            let queuedRedactions = this.toRedact.get(limitToRoomId);
+            if (queuedRedactions) {
+                await redact(queuedRedactions);
+                this.toRedact.delete(limitToRoomId);
```

Comment:
You should probably `delete` immediately. Otherwise, you risk race conditions in case a new redaction is added while you're `await`-ing.